akvo/rsr/models/tree/model.py

```diff
@@ -0,0 +1,130 @@
+from functools import cached_property
+from typing import List, Optional
+from uuid import UUID, uuid4
+
+from django.db import models, transaction
+from django_ltree.fields import PathValue
+from django_ltree.models import TreeModel
+
+from .errors import NodesWillBeOrphaned, TreeWillBreak
+from .helpers import label_to_uuid, uuid_to_label
+from .managers import AkvoTreeManager, AkvoTreeQuerySet
+
+
+class AkvoTreeModel(TreeModel):
+    objects = AkvoTreeManager
+
+    uuid = models.UUIDField(editable=False, default=uuid4, unique=True)
+
+    class Meta:
+        abstract = True
+
+    @cached_property
+    def manager(self) -> AkvoTreeManager:
+        return type(self)._default_manager
+
+    @property
+    def has_ancestors(self) -> bool:
+        return len(self.path) > 1
+
+    def ancestors(self, with_self=True) -> AkvoTreeQuerySet:
+        """
+        Get ancestors sorted by closest to farthest
+
+        That means parent, then parent's parent, etc.
+        """
+        ancestors = super().ancestors().order_by("-path")
+        if not with_self:
+            return ancestors.exclude(id=self.id)
+        return ancestors
+
+    def parent(self) -> Optional["AkvoTreeModel"]:
+        if self.has_ancestors:
+            return self.ancestors(with_self=False).first()
+
+    def get_parent_uuid(self) -> Optional[UUID]:
+        if len(self.path) > 1:
+            return label_to_uuid(self.path[-2])
+
+    def get_root(self) -> "AkvoTreeModel":
+        """
+        Get the root node of a tree, which can be the node itself
+        """
+        if len(self.path) > 1:
+            return self.manager.get(uuid=label_to_uuid(self.path[0]))
+        else:
+            return self
+
+    def descendants(self, with_self=True):
+        """
+        All sub-nodes and their subnodes and so on
+
+        :param with_self: Include self in the result
+        """
+        return self.manager.descendants(self.path, with_self=with_self)
+
+    def set_parent(self, new_parent: "AkvoTreeModel", force=False, update_descendants=True):
```

Comment:
Does this signature make sense? Should the call to `_update_descendant_parents` be moved out and left to the caller? What do you think @zuhdil ?

Reply:
I don't really get what's the need for `force` and `update_descendants`. In the previous codebase, we just set the parent, no need to force anything or tell the descendants to update themself. It will be great if the signature can be made simpler IMO.

Reply:
The problem is that because the path to the node is stored on the node, if the parent changes, all descendants won't be able to be found anymore. Updating a node with descendants will destroy the tree.

It's not possible to use signals as the object will already have the new path and won't be able to find the old descendants anymore without querying the DB first.

After thinking about it, I'll do what the old code did. The model method will just take the action with guarantees of tree stability (will add the warning to the docstring) and the usecase will take case of updating descendants, etc.